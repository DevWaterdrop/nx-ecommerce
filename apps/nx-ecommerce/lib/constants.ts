export const LOVELY_ERROR = 'I HATE TYPESCRIPT+STRAPI+GRAPHQL';
export const MAXIMUM_IN_CART = 16;

export enum PRICE_FILTER {
  ZERO_TWENTY = 'zt',
  TWENTY_FORTY = 'tf',
  FORTY_EIGHTY = 'fe',
  EIGHTY_PLUS = 'ep',
}

export enum SORT_FILTER {
  PRICE_LOW_HIGH = 'plh',
  PRICE_HIGH_LOW = 'phl',
  NAME = 'name',
}

export const PRICES = [
  { value: PRICE_FILTER.ZERO_TWENTY, label: '0.00$ - 19.99$' },
  { value: PRICE_FILTER.TWENTY_FORTY, label: '20.00$ - 39.99$' },
  { value: PRICE_FILTER.FORTY_EIGHTY, label: '40.00$ - 79.99$' },
  { value: PRICE_FILTER.EIGHTY_PLUS, label: '80.00$+' },
];

export const SORT = [
  { value: SORT_FILTER.PRICE_LOW_HIGH, label: 'Price: low to high' },
  { value: SORT_FILTER.PRICE_HIGH_LOW, label: 'Price: high to low' },
  { value: SORT_FILTER.NAME, label: 'Name' },
];
