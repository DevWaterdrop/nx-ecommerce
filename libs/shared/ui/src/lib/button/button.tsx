import clsx from 'clsx';
import Link from 'next/link';
import { useMemo } from 'react';

type ConditionalProps =
  | {
      tag: 'a';
      elProps?: JSX.IntrinsicElements['a'] & {
        href: NonNullable<JSX.IntrinsicElements['a']['href']>;
      };
    }
  | {
      tag: 'button';
      elProps?: JSX.IntrinsicElements['button'];
    };

export type ButtonProps = ConditionalProps & {
  size?: typeof SIZES[number];
  type?: typeof TYPES[number];
  round?: typeof ROUNDS[number];
  basic?: boolean;
  classes?: string;
  disabled?: boolean;
  children: React.ReactNode;
};

const SIZES = ['parent', 'base', 'lg', 'sm'] as const;
const TYPES = [
  'primary',
  'transparent',
  'black',
  'transparent-border',
] as const;
const ROUNDS = ['full', 'sm', 'base'] as const;

export const Button: React.FC<ButtonProps> = (props) => {
  const {
    tag,
    elProps,
    basic = false,
    size = 'base',
    type = 'primary',
    round = 'full',
    classes,
    disabled,
    children,
  } = props;

  const className = useMemo(
    () =>
      basic
        ? clsx(
            'inline-block',
            'cursor-pointer',
            'transform-gpu transition-all',
            'active:scale-92',
            // DISABLED
            disabled && 'opacity-50 pointer-events-none',
            classes
          )
        : clsx(
            'inline-block',
            'cursor-pointer',
            'transform-gpu transition-all',
            'active:scale-92',
            // ROUND
            round === 'full' && 'rounded-full',
            round === 'sm' && 'rounded-lg',
            round === 'base' && 'rounded-xl',
            // SIZE
            size === 'parent' && 'w-full h-full',
            size === 'sm' && 'p-2',
            size === 'base' && 'p-3',
            size === 'lg' && 'p-4',
            // TYPES
            type === 'primary' &&
              'text-white bg-sky-600 hover:(bg-sky-700) active:(bg-sky-800)',
            type === 'transparent' &&
              'hover:(bg-gray-100) active:(bg-gray-300)',
            type === 'black' && 'bg-black text-white hover:(bg-dark-400)',
            type === 'transparent-border' &&
              'hover:(bg-gray-100 border-black) active:(bg-gray-300) border border-gray-300',
            // DISABLED
            disabled && 'opacity-50 pointer-events-none',
            classes
          ),
    [type, size, round, classes, disabled, basic]
  );

  if (tag === 'a') {
    return (
      <Link href={elProps?.href || '/'}>
        <a className={className} {...elProps}>
          {children}
        </a>
      </Link>
    );
  }

  return (
    <button className={className} {...elProps}>
      {children}
    </button>
  );
};
