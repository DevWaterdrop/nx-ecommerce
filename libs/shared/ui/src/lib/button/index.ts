import { Button, ButtonProps } from './button';

export { Button };

export type { ButtonProps };
