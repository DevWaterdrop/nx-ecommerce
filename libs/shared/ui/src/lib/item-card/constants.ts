import { ItemCardProps } from './item-card';
import flowerJPEG from '../../props/flower.jpeg';
import roomJPEG from '../../props/room.jpeg';

const ITEM: Required<ItemCardProps['item']> = {
  price: 65,
  name: 'Lovely flower',
  id: '#',
  images: [flowerJPEG, roomJPEG],
  smallDescription: 'For you, for me, for us',
  tag: 'new',
};

export const DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleClick: () => console.log('clicked'),
};
