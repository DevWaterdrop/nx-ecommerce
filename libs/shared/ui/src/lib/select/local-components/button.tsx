import clsx from 'clsx';
import Image from 'next/image';
import type { Item, ItemMap } from '../select';
import chevronDownSVG from '../../../assets/icons/chevron-down.svg';

interface Props {
  selected: Item | ItemMap | null;
  placeholder: string;
  open: boolean;
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;
}

export const Button: React.FC<Props> = (props) => {
  const { selected, placeholder, open, setOpen } = props;

  const handleClick = () => void setOpen((prev) => !prev);

  return (
    <button
      className={clsx(
        'py-2.5 px-4',
        'flex items-end gap-2',
        'bg-gray-100',
        'text-sm font-medium',
        'rounded-full border border-transparent',
        'sm:(px-6)',
        (selected instanceof Map ? selected.size > 0 : selected !== null) &&
          'border-black'
      )}
      type="button"
      onClick={handleClick}
    >
      {placeholder}
      <div
        className={clsx(
          'h-4.5 w-4.5 relative',
          'transition-transform transform-gpu',
          open && 'rotate-180'
        )}
        aria-hidden="true"
      >
        <Image width={18} height={18} src={chevronDownSVG} alt="" />
      </div>
    </button>
  );
};
