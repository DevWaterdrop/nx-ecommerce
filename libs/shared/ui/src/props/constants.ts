import { ButtonProps } from '../lib/button';
import { CartItemProps } from '../lib/cart-item';
import { CartSidebarProps } from '../lib/cart-sidebar';
import { CategoryCardProps } from '../lib/category-card';
import { IconProps } from '../lib/icon/icon';
import { ItemCardProps } from '../lib/item-card';
import { NavButtonProps } from '../lib/nav-button';
import { OrdersTableProps } from '../lib/orders-table';
import { ProductGalleryProps } from '../lib/product-gallery';
import { ProductInformationProps } from '../lib/product-information';
import { ProductSidebarProps } from '../lib/product-sidebar';
import { SearchbarProps } from '../lib/searchbar';
import { SearchbarItemProps } from '../lib/searchbar-item';
import { SelectProps } from '../lib/select';
import flowerJPEG from './flower.jpeg';
import roomJPEG from './room.jpeg';

export const ITEM: ItemCardProps['item'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
};

export const ITEM_SECOND: ItemCardProps['item'] = {
  slug: 'roomy',
  price: 50,
  name: 'Roomy',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'Hmmm, nice vase',
  description: 'Vaaaase',
};

export const ITEM_CARD_DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleCartClick: () => console.log('cart clicked'),
  handleAmountCartClick: () => console.log('amount cart clicked'),
  handleFavoriteClick: () => console.log('favorite clicked'),
};

export const CATEGORY_CARD_DEFAULT_PROPS: CategoryCardProps = {
  category: {
    smallDescription:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
    image: {
      data: {
        attributes: {
          url: roomJPEG,
        },
      },
    },
    slug: '1',
  },
};

export const NAV_BUTTON_DEFAULT_PROPS: NavButtonProps = {
  type: 'heart',
};

export const ICON_DEFAULT_PROPS: IconProps = {
  icon: 'cart',
};

export const BUTTON_DEFAULT_PROPS: ButtonProps = {
  children: 'hello',
  tag: 'button',
  size: 'base',
  type: 'primary',
  round: 'full',
};

export const SEARCHBAR_ITEM: SearchbarProps['items'][number] = {
  id: '111',
  attributes: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
};

export const SEARCHBAR_ITEM_DEFAULT_PROPS: SearchbarItemProps = {
  item: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
  handleClick: () => void 1,
};

export const SELECT_ITEM = {
  label: 'Label1',
  value: 'value1',
};

export const SELECT_DEFAULT_PROPS: Required<
  Omit<SelectProps, 'placeholder' | 'multiple'>
> = {
  items: [SELECT_ITEM],
  initial: ['value1'],
  onChange: () => console.log('changed'),
};

export const PRODUCT: ProductSidebarProps['product'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
  information: [
    {
      title: 'Title1',
      content: 'content1',
    },
  ],
};

export const PRODUCT_SIDEBAR_DEFAULT_PROPS: ProductSidebarProps = {
  product: PRODUCT,
  handleCartClick: () => {
    console.log('cart');
  },
  handleRemoveCartClick: () => {
    console.log('remove cart');
  },
  handleFavoriteClick: () => {
    console.log('favorite');
  },
  cartValue: new Map([]),
  favoritesValue: new Set('lovely-flower'),
};

export const PRODUCT_INFORMATION_DEFAULT_PROPS: ProductInformationProps = {
  information: [...Array(10).keys()].map((_value, index) => ({
    title: `Title${index + 1}`,
    content: `content${index + 1}`,
  })),
};

export const PRODUCT_GALLERY_DEFAULT_PROPS: ProductGalleryProps = {
  images: [
    {
      attributes: {
        url: flowerJPEG,
      },
    },
    {
      attributes: {
        url: roomJPEG,
      },
    },
  ],
};

const PRODUCT_ORDER_TABLE: OrdersTableProps['orders'][number]['items'][number]['product'] =
  {
    name: 'Flower',
    slug: 'slug-flower',
    description: '',
    smallDescription: '',
    price: 50,
  };

export const ORDERS_TABLE_DEFAULT_PROPS: OrdersTableProps = {
  orders: [
    {
      id: 'id-1',
      amount: 50,
      items: [{ id: '1', amount: 1, product: PRODUCT_ORDER_TABLE }],
    },
    {
      id: 'id-2',
      amount: 100,
      items: [{ id: '2', amount: 2, product: PRODUCT_ORDER_TABLE }],
    },
  ],
};

export const CART_SIDEBAR_DEFAULT_PROPS: CartSidebarProps = {
  cartTotal: 100,
  isUser: true,
  handleCheckoutClick: async () => {
    console.log('checkout');
  },
  isLoading: false,
};

export const CART_ITEM_DEFAULT_PROPS: CartItemProps = {
  inCartAmount: 1,
  item: PRODUCT,
  handleCartMinusClick: () => {
    console.log('cart minus');
  },
  handleCartPlusClick: () => {
    console.log('cart plus');
  },
  handleRemoveFromCart: () => {
    console.log('cart remove');
  },
  toggleFavorite: () => {
    console.log('favorite');
  },
};
