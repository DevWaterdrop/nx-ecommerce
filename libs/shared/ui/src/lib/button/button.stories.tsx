import { BUTTON_DEFAULT_PROPS } from '../../props/constants';
import { Story, Meta } from '@storybook/react';
import { Button, ButtonProps } from './button';

export default {
  component: Button,
  title: 'Components/Button',
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = BUTTON_DEFAULT_PROPS;
