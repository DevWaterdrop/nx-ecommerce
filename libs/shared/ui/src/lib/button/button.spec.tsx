import { BUTTON_DEFAULT_PROPS } from '../../props/constants';
import { render, RenderResult, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { Button } from './button';

describe('Button', () => {
  let renderResult: RenderResult;
  const props = BUTTON_DEFAULT_PROPS;

  beforeEach(() => {
    renderResult = render(<Button {...props} />);
  });

  it('should render successfully', () => {
    const { baseElement } = renderResult;

    const tree = renderer.create(<Button {...props} />).toJSON();
    expect(tree).toMatchSnapshot();

    expect(baseElement).toBeTruthy();
  });

  it('should render as "a" or "button" tag', () => {
    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(screen.queryByRole('link')).toBeFalsy();

    renderResult.rerender(<Button tag="a">test</Button>);

    expect(screen.getByText('test').closest('a')).toBeInTheDocument();
    expect(screen.queryByRole('button')).toBeFalsy();
  });
});
