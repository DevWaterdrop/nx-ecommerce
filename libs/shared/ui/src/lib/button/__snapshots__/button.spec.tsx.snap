// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render successfully 1`] = `
<button
  className="inline-block cursor-pointer transform-gpu transition-all active:scale-92 rounded-full p-3 text-white bg-sky-600 hover:(bg-sky-700) active:(bg-sky-800)"
>
  hello
</button>
`;
